server/src-lib/Hasura/RQL/DDL/Remote/Validate.hs

```diff
@@ -0,0 +1,201 @@
+     HashMap QualifiedTable TableInfo
+  -> HashMap G.Name InpValInfo
+  -> HashMap G.Name G.Value
+  -> HashMap G.Variable FieldInfo
+  -> Validation (NonEmpty ValidationError) ()
+validateRemoteArguments tables expectedArguments providedArguments permittedVariables = do
+  traverse validateProvided (HM.toList providedArguments)
+  traverse validateExpected (HM.toList expectedArguments)
+  pure ()
+  where
+    validateProvided (providedKey, providedValue) =
+      case HM.lookup providedKey expectedArguments of
+        Nothing -> Failure (pure (NoSuchArgumentForRemote providedKey))
+        Just expectedType ->
+          validateType
+            tables
+            permittedVariables
+            providedValue
+            (_iviType expectedType)
+    validateExpected (expectedKey, expectedInpValInfo) =
+      case _iviDefVal expectedInpValInfo of
+        Just {} -> pure ()
+        Nothing ->
+          case HM.lookup expectedKey providedArguments of
+            Nothing -> Failure (pure (MissingRequiredArgument expectedKey))
+            Just {} -> pure ()
+
+-- | Validate a value against a type.
+validateType ::
+     HashMap QualifiedTable TableInfo
+  -> HashMap G.Variable FieldInfo
+  -> G.Value
+  -> G.GType
+  -> Validation (NonEmpty ValidationError) ()
+validateType tables permittedVariables value (gtypeToEither -> expectedType) =
+  bindValidation
+    (valueType tables permittedVariables value)
+    (\actualType ->
+       when (actualType /= expectedType) (Failure (pure InvalidType)))
+
+-- | Produce the type of a value.
+valueType ::
+     HashMap QualifiedTable TableInfo
+  -> HashMap G.Variable FieldInfo
+  -> G.Value
+  -> Validation (NonEmpty ValidationError) (Either G.NamedType G.ListType)
+valueType tables permittedVariables =
+  \case
+    G.VVariable variable ->
+      case HM.lookup variable permittedVariables of
+        Nothing -> Failure (pure (InvalidVariable variable permittedVariables))
+        Just fieldInfo -> fmap Left (fieldInfoToNamedType tables fieldInfo)
+    G.VInt {} -> pure (Left (mkScalarTy PGInteger))
+    G.VFloat {} -> pure (Left (mkScalarTy PGFloat))
+    G.VBoolean {} -> pure (Left (mkScalarTy PGBoolean))
+    G.VNull -> Failure (pure NullNotAllowedHere)
+    G.VString {} -> pure (Left (mkScalarTy PGText))
+    (G.VEnum _) -> error "TODO: G.VEnum: What do we do here?"
+    (G.VList _) -> error "TODO: G.VList: What do we do here?"
+    (G.VObject _) -> error "TODO: G.VObject: What do we do here?"
+
+-- | Convert a field info to a named type, if possible.
+fieldInfoToNamedType ::
+     HashMap QualifiedTable TableInfo
+  -> FieldInfo
+  -> Validation (NonEmpty ValidationError) G.NamedType
+fieldInfoToNamedType tables =
+  \case
+    FIColumn pgColInfo -> pure (mkScalarTy (pgiType pgColInfo))
+    FIRelationship relInfo ->
+      case HM.lookup (riRTable relInfo) tables of
+        Nothing -> Failure (pure (InvalidForeignTable relInfo))
+        Just _tableInfo ->
+          Failure
```

Comment:
Which case is this? Are we not allowing remote relationships between local relationships and remotes?

We should discuss this.

Reply:
We should discuss it, yep. Basically the case would be:

```haskell
local_table {
  remote_field1
  remote_field2
}
```

where `remote_field2` depends on `remote_field1`.

We can support this, by querying the results from hasura, then remote1, then remote2, which would preclude parallelism in that case.

Reply:
No, we don't need to support this.

Reply:
Cool. :+1: